File: 0494-target-sum/0494-target-sum.py
class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        n = len(nums)
        @cache
        def helper(i, s):
            if i >= n:
                if s == target:
                    return 1
                return 0 
                
            plus = helper(i + 1, s + nums[i])
            neg  = helper(i + 1, s - nums[i])
                
            return plus + neg
        return helper(0, 0)

File: 0494-target-sum/0494-target-sum.cpp
class Solution {
public:
    int findTargetSumWays(vector<int>& nums, int target) {
        int count = 0;
        helper(nums, 0, 0, &count, target);
        return count;
    }
    
    void helper(vector<int> &nums, int sum, int i, int* count, int target) {
        if (i == nums.size()) {
            if (sum == target) {
                (*count)++;
            }    
        }
        else {
            helper(nums, sum + nums[i], i + 1, count, target);
            helper(nums, sum - nums[i], i + 1, count, target);
        }
        
        
    }
};